### DISCUSSION TITLE
[Flutter/Dart] Support Dart: Use Dart FFI

### DISCUSSION THREAD

**Discussion by hhstore** [trustwallet/wallet-core#1264] [Ideas & Feature request]
# Support Dart: 

- since i want to use `wallet-core` in dart. 
- so i try to use `dart:ffi` to make it. 
- but i am not quite familiar with `wallet-core`. 
- anyone who wants to help me? or any suggestions on where to start first?
- i would really appreciate it.

## my fork: 

- https://github.com/better-dart/trustwallet-core/tree/dart/ffi-base-on-v2.3.9
- dev-branch: `dart/ffi-base-on-v2.3.9`
- i will add `dart:ffi` to support dart here.

## dart ffi: 

- https://dart.dev/guides/libraries/c-interop
- https://flutter.dev/docs/development/platform-integration/c-interop

### libs: 

- [x] https://github.com/dart-lang/ffi
- [x] https://github.com/dart-lang/ffigen
    - a tool to auto gen binding codes.

### usage: c++ with ffi 

- https://github.com/dart-lang/tflite_native

![image](https://user-images.githubusercontent.com/3252130/101440225-3a689a00-3951-11eb-9ba5-46f2574e0581.png)

# Maybe these are the things I need?

## docs: 

- binding: https://developer.trustwallet.com/wallet-core/developing-the-library/building

- ios: https://developer.trustwallet.com/wallet-core/integration-guide/ios-guide

```
 pod 'TrustWalletCore'

```

- android: https://developer.trustwallet.com/wallet-core/integration-guide/android-guide

> jni load: 

```java
init {
    System.loadLibrary("TrustWalletCore")
}

```

- go: https://developer.trustwallet.com/wallet-core/integration-guide/server-side

## *.h

- https://github.com/trustwallet/wallet-core/blob/81a5b234f6/include/TrustWalletCore/TWHDWallet.h

## jni: 

- https://github.com/trustwallet/wallet-core/blob/81a5b234f6/jni/cpp/TWJNI.h

## gen proto ? 

- https://github.com/trustwallet/wallet-core/blob/e8ff9c7106/tools/generate-files

- pb: https://github.com/trustwallet/wallet-core/blob/e8ff9c7106/src/proto/Ethereum.proto

## iOS: 

- https://github.com/trustwallet/wallet-core/blob/81a5b234f6/tools/ios-build

## Android: 

- https://github.com/trustwallet/wallet-core/blob/81a5b234f6/tools/android-build

## code gen: 

- https://github.com/trustwallet/wallet-core/blob/99af2f62cd/codegen/lib/code_generator.rb

**Comment by hhstore** [comment]
## The latest progress:

- https://github.com/better-dart/wallet-core/blob/dart%2Fffi-base-on-v2.3.9/dart/README.md
- the `dart binding file` has been successfully generated, and I am testing it. 

### pubspec.yaml

- https://github.com/better-dart/wallet-core/blob/dart%2Fffi-base-on-v2.3.9/dart/pubspec.yaml

### dart binding file: 

- https://github.com/better-dart/wallet-core/blob/dart%2Fffi-base-on-v2.3.9/dart/lib/wallet_core_bindings.dart#L7
- this is the generate binding codes.

**Comment by hewigovens** [comment]
Are you building dart bindings on top of existing iOS/android lib?

**Comment by hhstore** [comment]
> Are you building dart bindings on top of existing iOS/android lib?

- i think i need  `iOS/Android` dynamic library. 
- but i not sure, if `Android wallet-core.aar` is what i needed? or i should build a dynamic library by myself?
- from this doc, https://developer.trustwallet.com/wallet-core/developing-the-library/releasing
    - https://bintray.com/trust/wallet-core/com.trustwallet.wallet-core
- i find `android` wallet-core.aar file. 
- i still working on how to add `android wallet-core.aar` to my flutter project and call it via `dart binding codes`, since the docs related to `dart:ffi` are really rare.
- in this process, I encountered some problems, some errors. i'm working on it.

## about dart ffi usage: 


- https://flutter.dev/docs/development/platform-integration/c-interop#step-3-load-the-code-using-the-ffi-library


![image](https://user-images.githubusercontent.com/3252130/101438114-941a9580-394c-11eb-8a0a-1a763b70dd3a.png)

## rust + ffi + flutter: 


- [x] i find a rust ffi binding usage here: https://github.com/Hentioe/mikack-ffi
    - https://github.com/Hentioe/mikack-ffi/blob/master/Makefile#L42
- [x] this a rust ffi usage, use for a flutter project. i think i'm going to do something like this. 

![image](https://user-images.githubusercontent.com/3252130/101438341-215dea00-394d-11eb-9dc2-af3dca54d8dd.png)

**Comment by hewigovens** [comment]
Hey, what's status of dart binding? I think you can skip building process and unzip aar from here https://bintray.com/trust/wallet-core/com.trustwallet.wallet-core/2.4.1

**Comment by jayateertha043** [comment]
Hey, 
I was also looking for the same integration ,if you managed to integrate trust wallet into flutter kindly let me know

Thanks and regards
Jayateertha G

**Comment by jayateertha043** [comment]
Can't we just put the aar files into android/lib/ folder and interact with them using method channel ?

**Comment by hhstore** [comment]
@hewigovens 


- It's not going well. 
- I encountered some native-related problems, which I can't solve at the moment.
- Due to time constraints, I am temporarily using the `dart native library` instead.
- However, I will still consider replacing it with `trustwallet-core`.
- I would still try the `FFI` or `channel method` to bind the `trustwallet-core`.

## Follow up:

- The latest developments will be synchronized in the following 2 projects.


### use FFI: 

- https://github.com/better-dart/trustwallet-core


### use Platform Channels: 


- https://github.com/better-dart/flutter_trust_wallet_core
- channel method

**Comment by hhstore** [comment]
@jayateertha043 

- Using the `channel method` is a good choice. 
- To use the channel method, you need to write Android/iOS `native codes`, but I'm not familiar with native. 
- If you are familiar with native coding, you can try it.

**Comment by memtopia** [comment]
I'm impl a Flutter plugin to doing this, most method are wrap, will keep updating, welcome pull request 

[https://github.com/weishirongzhen/flutter_trust_wallet_core](https://github.com/weishirongzhen/flutter_trust_wallet_core)

**Reply by hhstore** [reply]
> I'm impl a Flutter plugin to doing this, most method are wrap, will keep updating, welcome pull request
> 
> https://github.com/weishirongzhen/flutter_trust_wallet_core

Wow, thanks a lot. I will try to use it sometime.

**Reply by hhstore** [reply]
Here are two suggestions:
- flutter 2.x: `null safety` support. maybe I can help you with that. 😄
- publish the package to https://pub.dev/

**Reply by memtopia** [reply]
About null safety, it's already supported, see yaml dark sdk >=2.12.
And pub.dev,  cause pub.dev require plugin size <=50M, currently the android aar and ios xcframework are include in this repo, so it's unable to publish to pub.
btw, android aar need's to modify CMakeLists to rebuild to expose c api for dart side, in the future, i think native lib needs to be separate.

**Reply by hewigovens** [reply]
nice work, I checked your repo, it seems built binary are checked in directly into git (not that good), is it possible to use official binaries? we have plan to release dynamic iOS frameworks

**Reply by memtopia** [reply]
You can use ios lib from official, but android lib need's to be set preset visibility to default and rebuild from source, otherwise ffi can not lookup function from android side.

**Reply by hhstore** [reply]
My mistake. just noticed the `flutter` version. already supported `null safety` is great. 

Is there any way to solve the `android aar` problem?

**Reply by hhstore** [reply]
I'm running the `example` on my iPhone, and it seems to be working fine with `flutter.2.2.3`. 

I will take the time to migrate my project (previously using web3dart), thank you very much.

**Reply by hhstore** [reply]
![image](https://user-images.githubusercontent.com/3252130/130167368-ea765bb0-380e-4de3-82f8-ef8d6177bbf0.png)

**Reply by hhstore** [reply]
https://discord.com/invite/MnDA9pfWAW

btw, I have created a discord technical discussion group, feel free to join and discuss flutter related technologies.

**Comment by hhstore** [comment]
## a PR here: 


- https://github.com/weishirongzhen/flutter_trust_wallet_core/pull/13 
- I replaced the `iOS/Android` framework dependencies with the `official sources`.
- I submitted a PR, solving the problem of publishing to pub.dev. 
I tested it locally for iOS and it works fine, but there are some issues with the Android dependency changes. 
- However, you can choose to keep the `android aar`, and you can submit https://pub.dev/ directly
- Or, you can choose to fix the Android error and then publish it. 
- I'm not familiar with native, I can't fix this error.  I need some help. The error message is here. 
- https://github.com/weishirongzhen/flutter_trust_wallet_core/issues/12



### Android aar file: 

- https://github.com/weishirongzhen/flutter_trust_wallet_core/blob/main/android/libs/trustwalletcore.aar
- just  `12.5MB`. it's ok for publish to https://pub.dev/  now.

### iOS frameworks: 

- https://github.com/weishirongzhen/flutter_trust_wallet_core/blob/main/ios/Frameworks/WalletCore.xcframework/ios-arm64/WalletCore.framework/WalletCore
- this is so big. 63.8MB, remove this is safety. 
- https://github.com/weishirongzhen/flutter_trust_wallet_core/blob/main/ios/Frameworks/WalletCore.xcframework/ios-arm64_x86_64-maccatalyst/WalletCore.framework/WalletCore



@weishirongzhen @hewigovens

**Comment by hhstore** [comment]
- btw, I suggest merging it [flutter_trust_wallet_core](https://github.com/weishirongzhen/flutter_trust_wallet_core) into the official repo. 
- It is convenient for more people to use it.

**Reply by memtopia** [reply]
There still lots of things todo, especially AnySigner and  Protobuf side. Still lack of Complete test case, code check，and so on.... After all this done, then I'll consider  make a pull request to this repo.

**Reply by hewigovens** [reply]
Added to readme: https://github.com/trustwallet/wallet-core#community

**Comment by baahkusi** [comment]
Impressive work @weishirongzhen . Also started a trustwallet-core flutter plugin using `methodChannels` @hhstore .  The project is at a much younger stage as a proof of concept but starting to get more attention.

https://github.com/EjaraApp/trustdart

**Comment by Ce1ling** [comment]
Hey 👋, do you guys have any updates on this matter? I think it’s really important to make it support Flutter.

**Comment by Ruyeduardo** [comment]
Banco Beneficiário: Banco Topazio S/A | Código SWIFT: TOPZBRRSXXX | IBAN:
BR9407679404000000598183965C1 | Titular da Conta: Ruy Eduardo Barbosa De
Assuncao Maia | Endereço: Rua 18 de Novembro, 273 - Porto Alegre - RS,
90240-040 /// Banco Intermediário: Standard Chartered Bank | Código SWIFT:
SCBLUS33 | Titular da Conta: Banco Topazio S/A | Número da Conta:
3544026839001 | Endereço: New York, USA

Em qua., 28 de fev. de 2024 00:42, L1en ***@***.***> escreveu:
